pkg/servicenameresolutiondetector/coredns/detector_test.go

```diff
@@ -0,0 +1,547 @@
+/*
+Copyright 2024 The Karmada Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package coredns
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+	componentbaseconfig "k8s.io/component-base/config"
+	"k8s.io/klog/v2"
+
+	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
+	karmadafake "github.com/karmada-io/karmada/pkg/generated/clientset/versioned/fake"
+)
+
+type mockConditionStore struct {
+	conditions []metav1.Condition
+}
+
+func (m *mockConditionStore) ListAll() ([]metav1.Condition, error) {
+	return m.conditions, nil
+}
+
+func (m *mockConditionStore) Load(_ string) (*metav1.Condition, error) {
+	return nil, nil
+}
+
+func (m *mockConditionStore) Store(_ string, _ *metav1.Condition) error {
+	return nil
+}
+
+func TestNewCorednsDetector(t *testing.T) {
+	tests := []struct {
+		name        string
+		hostName    string
+		clusterName string
+		expectedErr bool
+	}{
+		{
+			name:        "Valid detector creation",
+			hostName:    "test-host",
+			clusterName: "test-cluster",
+			expectedErr: false,
+		},
+		{
+			name:        "Empty host name",
+			hostName:    "",
+			clusterName: "test-cluster",
+			expectedErr: false,
+		},
+		{
+			name:        "Empty cluster name",
+			hostName:    "test-host",
+			clusterName: "",
+			expectedErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			memberClusterClient := fake.NewSimpleClientset()
+			karmadaClient := karmadafake.NewSimpleClientset()
+			informerFactory := informers.NewSharedInformerFactory(memberClusterClient, 0)
+
+			cfg := &Config{
+				PeriodSeconds:    time.Second,
+				SuccessThreshold: time.Minute,
+				FailureThreshold: time.Minute,
+				StaleThreshold:   time.Minute,
+			}
+
+			leConfig := componentbaseconfig.LeaderElectionConfiguration{
+				LeaderElect: false,
```

Comment:
Can we use `leConfig` as one of the test's input conditions to test scenarios where detector creation fails?

Reply:
Yes, i'll add it in input conditions.